### PR TITLE
add C++ namespacing logic to ctypes

### PR DESCRIPTION
**Pull request by bdhirsh**
Added namespaces to all of the `CTypes` printed in the codegen. This is pretty much required if we want to use codegen externally, since we can no longer assume that we're inside of the `at::` namespace.

Important changes are in `types.py`.

How do we add the notion of namespaces to C++ types without people having to write "at::Tensor" everywhere? Before this PR, `CType` held a raw string representing the type, i.e. `BaseCType("Tensor", binds)`. This PR introduces a set of singleton base C++ types in `types.py`, that know how to print their namespace. Instead, we'd write `BaseCType(tensorT, binds)`, where printing `tensorT` will properly print out "at::Tensor".

This also means that you can't create arbitrary `CTypes`. If we need a new C++ type in the codegen, we need to add it to the list in `types.py`.

One blip in the design: we don't want to change `RegistrationDeclarations.yaml`, since that'll break external backends that ingest it. I added separate functions to display types without the namespace that are used to create RegistrationDeclarations.yaml`. With an external codegen API though, we can eventually kill it :)

I also didn't realize until this PR that `Declarations.yaml` is still directly in use, by some python/autograd codegen. Rather than keep that yaml byte-for-byte compatible, I just updated the callsites in the autograd codegen to work with namespaces. In the NEXT pr, I try to clean up some of the autograd codegen to stop using raw strings to match against C++ types.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55796 remove bridge API from codegen
* #55795 add _to_cpu() operator
* #55050 [EXPERIMENTAL, WIP] generate xla codegen in-tree
* #55334 split out NamedCType from CType. Remove direct string comparison from autograd codegen
* **#55047 add C++ namespacing logic to ctypes**
* #55046 Return a CType instead of a string for returns, beef up CType

Differential Revision: [D27708349](https://our.internmc.facebook.com/intern/diff/D27708349)

### CODE REVIEW

File: tools/codegen/api/types.py
Comment (bdhirsh) [inline tools/codegen/api/types.py:60]: I couldn't use `frozen=True` here, since I have a custom `__init__` function :/ added the `unsafe_hash` property instead so we can still use BaseCType as dictionary keys.

Comment (ezyang) [reply tools/codegen/api/types.py:55]: You can use `frozen=True`, you just have to be a little clever about how you do attribute setting. https://stackoverflow.com/questions/57893902/how-can-i-set-an-attribute-in-a-frozen-dataclass-custom-init-method

But I would examine more closely why you want this to happen in the constructor, since you can always define a helper class method for constructing frozen dataclasses. Maybe that is warranted here?

File: tools/codegen/api/types.py
Comment (ezyang) [inline tools/codegen/api/types.py:184]: The "remove" here is a bit too verb-y for me, it implies you are removing the namespaces from the CType (which you are not). Also, since this is a hack for a very specific situation (registration declarations), it is better to name the method off of how it is intended to be used, not what it does (because the functionality of a hack doesn't always make sense, but the motivation is always useful to know)

File: tools/codegen/api/types.py
Comment (ezyang) [inline tools/codegen/api/types.py:38]: One thing that's a little goofy is that for half of these types, `at::` isn't even the canonical namespace; they were originally defined in `c10::`. It's probably OK if the codegen studiously pretends that c10:: namespace doesn't exist though

File: tools/codegen/api/types.py
Comment (ezyang) [reply tools/codegen/api/types.py:184]: The name isn't even accurate, you don't remove *all* namespaces, you only remove the `at::` namespace

File: tools/codegen/api/types.py
Comment (ezyang) [inline tools/codegen/api/types.py:32]: Use of an Enum here seems a bit questionable, because you're using enum values that have double colons in their names, guaranteeing that you cannot reference enum values as BaseCppType.std::string, for example

Comment (ezyang) [reply tools/codegen/api/types.py:32]: A more common way to model this sort of situation is to define some sort of structure that records the entity name and namespace, and then define a bunch of singleton objects for each entity variant.

```
@dataclass(frozen=True)
class BaseCppType:
  ns: Optional[str]
  name: str

Int64 = BaseCppType(None, "int64_t")
Tensor = BaseCppType("at", "Tensor")
```

You stop being able to pattern match over BaseCppType but you never make use of this capability anyway.

**Comment by ezyang** [comment]
I'm a bit torn; I half want to make you get the autograd codegen more typed so that the namespace change here would have been a no-op, but I also understand this is kind of a long detour on top of the detour you're already on here. I don't know how long it took for you to find all these sites, since the typechecker wouldn't have been any help here.

File: tools/codegen/api/types.py
Comment (bdhirsh) [reply tools/codegen/api/types.py:32]: Using an actual class instead of an enum to capture the namespace/name pair sounds good, and I like the idea of a set of singletons that are "valid" instances of the type.

File: tools/codegen/api/types.py
Comment (bdhirsh) [reply tools/codegen/api/types.py:184]: fair point. It's not "remove namespaces", since we don't remove `std::`. , it's "make things fit the specific format of RegistrationDeclarations.yaml. How about `def cpp_type_registration_declarations()`.

**Comment by bdhirsh** [comment]
> I'm a bit torn; I half want to make you get the autograd codegen more typed so that the namespace change here would have been a no-op, but I also understand this is kind of a long detour on top of the detour you're already on here. I don't know how long it took for you to find all these sites, since the typechecker wouldn't have been any help here.

The compiler caught all of the errors, and I had to map them to specific call-sites (wasn't too hard since I could just search for instances of Scalar/Tensor/TensorList).

It hopefully shouldn't be too much work to make a minor improvement to the autograd codegen- instead of comparing the types to string (which need to know about namespaces), I can just compare them to other types. I'll try to add that in a followup PR. At first glance, it'll require refining some of the autograd types, e.g. `SavedVariable` currently represents its type with a string.

The example that I have in my head: `var.type == 'at::Scalar'` will become `var.type == to_cpp_type('Scalar')`, where the two objects being compared are both instances of `BaseCppType` instead of raw strings.

File: tools/codegen/api/types.py
Comment (ezyang) [reply tools/codegen/api/types.py:38]: Byte for byte isn't a fundamentalist religion: you can violate it if it makes sense :) (Just more debugging out the other end!)

**Comment by ezyang** [comment]
FWIW, `to_cpp_type('Scalar')` is modestly worse than a real reference, mostly because it's less well typed (so less errors can be caught by mypy)

File: tools/codegen/api/types.py
Comment (bdhirsh) [reply tools/codegen/api/types.py:38]: yeah :p I've found it to be really helpful in uncovering diffs between my codegen output's current state -> target state. I could always just keep it byte-for-byte locally though until I'm satisfied, then remove pointless diffs before I merge.

**Comment by bdhirsh** [comment]
@ezyang I took your suggestion of using named singletons. I also left all of the `if .... == 'at::Tensor'` mess in this PR, and tried to save the improvement for the followup PR (next in the stack). I didn't get fully to comparing JIT types, but it compares against `CType`s now instead of raw strings

**Review by ezyang — APPROVED** [review]
Looks good, we can always change this later if need be. Thanks for shaving this one!

File: tools/codegen/api/types.py
Comment (bhosmer) [inline tools/codegen/api/types.py:92]: Just to confirm, this is the only namespace we('ll ever) need to strip?

File: tools/codegen/api/types.py
Comment (bhosmer) [inline tools/codegen/api/types.py:104]: I don't know how much appetite you have for tweaking this code, but if you added `strip_namespace` to `cpp_type()` alongside `strip_ref`, you could do away with all these extra methods...

**Review by bhosmer — APPROVED** [review]
I wouldn't hold up landing this on anything I'm suggesting below unless it's appealing to make the changes here, but I'd at least consider them for a followup.

File: tools/codegen/api/types.py
Comment (bdhirsh) [reply tools/codegen/api/types.py:32]: Yeah, I think you're totally right.

One way I see doing that is that I'd add more singletons for templated types to keep track of namespaces, i.e. `optionalT = BaseCppType('c10', 'optional')`.

That has the downside though of letting you create invalid CType's, i.e. it would be nice if mypy caught `BaseCType(optionalT)` as an invalid CType.

I think I'm going to try to land this and the last few PR's soon to avoid more merge conflicts, but I'll try to address this and the other comments in a later PR!

**Pull request closed without merging**
